# STA130 Week 01 Homework 

Please see the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) for the list of topics covered in this homework assignment, and a list of topics that might appear during ChatBot conversations which are "out of scope" for the purposes of this homework assignment (and hence can be safely ignored if encountered)

<details class="details-example"><summary style="color:blue"><u>Introduction</u></summary>

### Introduction
    
A reasonable characterization of STA130 Homework is that it simply defines a weekly reading comprehension assignment. 
Indeed, STA130 Homework essentially boils down to completing various understanding confirmation exercises oriented around coding and writing tasks.
However, rather than reading a textbook, STA130 Homework is based on ChatBots so students can interactively follow up to clarify questions or confusion that they may still have regarding learning objective assignments.

> Communication is a fundamental skill underlying statistics and data science, so STA130 Homework based on ChatBots helps practice effective two-way communication as part of a "realistic" dialogue activity supporting underlying conceptual understanding building. 

It will likely become increasingly tempting to rely on ChatBots to "do the work for you". But when you find yourself frustrated with a ChatBots inability to give you the results you're looking for, this is a "hint" that you've become overreliant on the ChatBots. Your objective should not be to have ChatBots "do the work for you", but to use ChatBots to help you build your understanding so you can efficiently leverage ChatBots (and other resources) to help you work more efficiently.<br><br>

</details>

<details class="details-example"><summary style="color:blue"><u>Instructions</u></summary>

### Instructions
    
1. Code and write all your answers (for both the "Prelecture" and "Postlecture" HW) in a python notebook (in code and markdown cells) 
    
> It is *suggested but not mandatory* that you complete the "Prelecture" HW prior to the Monday LEC since (a) all HW is due at the same time; but, (b) completing some of the HW early will mean better readiness for LEC and less of a "procrastentation cruch" towards the end of the week...
    
2. Paste summaries of your ChatBot sessions (including link(s) to chat log histories if you're using ChatGPT) within your notebook
    
> Create summaries of your ChatBot sessions by using concluding prompts such as "Please provide a summary of our exchanges here so I can submit them as a record of our interactions as part of a homework assignment" or, "Please provide me with the final working verson of the code that we created together"
    
3. Save your python jupyter notebook in your own account and "repo" on [github.com](github.com) and submit a link to that notebook though Quercus for assignment marking<br><br>

</details>

<details class="details-example"><summary style="color:blue"><u>Prompt Engineering?</u></summary>

### Prompt Engineering?    
    
The questions (as copy-pasted prompts) are designed to initialize appropriate ChatBot conversations which can be explored in the manner of an interactive and dynamic textbook; but, it is nonetheless **strongly recommendated** that your rephrase the questions in a way that you find natural to ensure a clear understanding of the question. Given sensible prompts the represent a question well, the two primary challenges observed to arise from ChatBots are 

1. conversations going beyond the intended scope of the material addressed by the question; and, 
2. unrecoverable confusion as a result of sequential layers logial inquiry that cannot be resolved. 

In the case of the former (1), adding constraints specifying the limits of considerations of interest tends to be helpful; whereas, the latter (2) is often the result of initial prompting that leads to poor developments in navigating the material, which are likely just best resolve by a "hard reset" with a new initial approach to prompting.  Indeed, this is exactly the behavior [hardcoded into copilot](https://answers.microsoft.com/en-us/bing/forum/all/is-this-even-normal/0b6dcab3-7d6c-4373-8efe-d74158af3c00)...

</details>

### Marking Rubric (which may award partial credit) 

- [0.1 points]: All relevant ChatBot summaries [including link(s) to chat log histories if you're using ChatGPT] are reported within the notebook
- [0.2 points]: Reasonable general definitions for Question "2.2"
- [0.2 points]: Demonstrated understanding regarding Question "4"
- [0.2 points]: A sensible justification for the choice in Question "7.4"
- [0.3 points]: Requested assessment of ChatBot versus google performance in Question "8"

## "Prelecture" HW [*completion prior to next LEC is suggested but not mandatory*]

#### 1. Pick one of the datasets from the ChatBot session(s) of the **TUT demo** (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> If your TA has not shared a relevant ChatBot session from their **TUT demo** through a piazza post and a Quercus announcement, the **TUT notebook** has links to example ChatBot sessions that you can use; or, ...
> 
> ```python
> # feel free to just use the following if you prefer...
> import pandas as pd
> url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
> df = pd.read_csv(url)
> df.isna().sum()
> ```
    
</details>


In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
df = pd.read_csv(url)
df.isna().sum()

#               0
Name            0
Type 1          0
Type 2        386
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64


#### 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a `pandas` DataFrame has, and then

1. use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,  
2. write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> A good initial prompt to start would be be something like
> - "I've downloaded a dataset about characters from animal crossings (from https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv), and I'd like to know what columns of information I have and how much data I have"
> 
> You can further reduce the scope of your inquiry with if needed with something like
> - "I've already downloaded the data and want to understand the size (or dimensions) of the dataset to start with"
> 
> *Some ChatBots can upload your data and do this for you; but, extended usage of this feature [likely requires a paid subscription](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/GPT/SLS/00006_gpt3p5_LoadDataPaywall.md); and, anyway, you need to run the code yourself rather than having a ChatBot do that for you; and, for STA130 we don't want a ChatBot to just do the analysis for us; rather, we instead want ChatBots to help us understand the steps we need to take to analyze the data; so,* **you DO NOT need to purchase an upgraded version of any ChatBots**
> - Free-tier level ChatBots like [GPT4o-mini](https://chat.openai.com/) or [Copilot](https://copilot.microsoft.com/) (which is partially based on [ChatGPT4.0](https://chat.openai.com/), and which you have access to through your UofT account) are sufficiently sophisticated and perfectly appropriate for the STA130 course
    
</details>


In [3]:
# Get the number of rows and columns
num_rows, num_columns = df.shape

print(f'Number of rows: {num_rows}')
print(f'Number of columns: {num_columns}')

Number of rows: 800
Number of columns: 12


Definition of variable vs. observation:

Variables are the columns in the dataframe, in other words, how you group each data. In my pokemon dataset, this would be the name, type1, type2, HP, attack, defense, etc.

Observations are the rows in the dataframe, in other words, the data within each topic/ "group". In my pokemon dataset, an example would be "bulbasaur", "ivysaur", "venusaur"... under the variable "name".

#### 3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Use your ChatBot session to help you create working examples of using  `df.describe()` and `df['column'].value_counts()` for your dataset (although note that the `.value_counts()` method is not really meant to be used for numeric variables, so if you dataset has only numeric variables, `.value_counts()` might not be particularly informative...)
>
> #### ChatBot Response Scope 
>     
> If prompts are not sufficiently focused you will likely get overly broad responses from the ChatBot, but you can always respond with subsequent refinement requests to appropriately limit the scope of the ChatBot responses to focus on addressing your actual content targets; so, 
> - an initially very general inquiry like, "I need help analyzing my data" will likely result in a ChatBot response suggesting a wide variety of approaches and techniques for summarizing your dataset; but, re-prompting the ChatBot with something like, "What's the simplest form of summarization of this dataset that I could do and how do I do it in Python?" or suggesting guidance using the specific summarization methods requested above will helpfully re-orient the ChatBot to your specific interests and needs
> 
> #### Jupyter Notebook Hints
> 
> Jupyter notebook printouts usaully don't show all of the data (when there's too much to show, like if `df.describe()` includes results for many columns), but the printouts just show enough of the data to give an idea of what the results are which is all we're looking for at the moment
> 
> - Consider dividing the code that ChatBot provides you into different jupyter notebook cells so that each cell concludes with a key printed result; the last line of code in a jupyter notebook cell will automatically print out in a formatted manner, so replacing something like `print(df.head())` with `df.head()` as the last line of a cell provides a sensible way to organize your code
> - The printout suggestions above are demonstrated in `STA130F24_CourseProject.ipynb` if looking at an example would be helpful to understand what they're getting at...
    
</details>

In [4]:
print(df.describe())

                #          HP      Attack     Defense     Sp. Atk     Sp. Def  \
count  800.000000  800.000000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750   69.258750   79.001250   73.842500   72.820000   71.902500   
std    208.343798   25.534669   32.457366   31.183501   32.722294   27.828916   
min      1.000000    1.000000    5.000000    5.000000   10.000000   20.000000   
25%    184.750000   50.000000   55.000000   50.000000   49.750000   50.000000   
50%    364.500000   65.000000   75.000000   70.000000   65.000000   70.000000   
75%    539.250000   80.000000  100.000000   90.000000   95.000000   90.000000   
max    721.000000  255.000000  190.000000  230.000000  194.000000  230.000000   

            Speed  Generation  
count  800.000000   800.00000  
mean    68.277500     3.32375  
std     29.060474     1.66129  
min      5.000000     1.00000  
25%     45.000000     2.00000  
50%     65.000000     3.00000  
75%     90.000000     5.00000  
max    180.000

In [5]:
#df['column'].value_counts() -> counts how many repeats for each observation given a certain variable/column.
df['Legendary'].value_counts()

Legendary
False    735
True      65
Name: count, dtype: int64

#### 4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, 
# explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by 
# `df.shape` and what is reported by `df.describe()` with respect to 
# (a) the number of columns it analyzes and (b) the values it reports in the "count" column<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> If the dataset you're using does not have (a) non-numeric variables and (b) missing values in numeric variables (e.g., the `"villagers.csv"` example above has only a single numeric variable `row_n` which has no missing values), instead download and use the [https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv](https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv)" data to answer this question  
>
> In (a) above, the "columns it analyzes" refers to the columns of the output of `df.describe()` which will only include "numeric" columns by default, but you can can see the names of all the columns in a dataset using `df.columns`; and, make sure `df.shape` is refering to the dataset you think it is... if you've loaded a different dataset it might not have been called `df`(!)
>
> **If you get any errors (for example related to column names), copy and paste them as a response to the ChatBot, and see if it can help you resove them by adding the suggested adjustments to your code and then reruning all your code to see if the changes have fixed the problem (and repeat this process as needed until the problems have been resolved).**
    
</details>

In [6]:
print(df.shape)

(800, 12)


In [7]:
df.isnull().sum()

#               0
Name            0
Type 1          0
Type 2        386
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

My data set has a) non-numeric variables and b) missing values in numeric values (386 missing values under variable "type 2", as shown by df.isnull().sum()).

df.shape returns (800,12) meaning  800 columns and 12 variables is reported— the entire size of the dataframe, regardless of whether they are numeric or not, or whether they have missing values or not.

df.describe () returns a count of 800 columns but 8 variables only, this is because df.describe() only doesn't take into account non-numeric variables. Since i have 4 non-numeric variables (name, type1, type2, and legendary), this makes sense.


#### 5. Use your ChatBot session to help understand the difference between the following and then 
#provide your own paraphrasing summarization of that difference

- an "attribute", such as `df.shape` which does not end with `()`
- and a "method", such as `df.describe()` which does end with `()` 

attributes directly accesses the information of an object and retrieve & return the information. it doesn't need a () because it doesn't recall any functions nor involve any computation.it holds the data and doesn't need any operations to access the data. 

method performs operation(s) on an object when called, such as computation. The () at the back indicates that it needs execution and arguments can be passed in in order for the operation to take place. It can be left empty, such as df.describe(), when you don't need to pass in any arguments. 


<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> The fact that a "method" such as `df.describe()` ends with `()` suggests that "methods" are essentially something that we would call a "function" in programming language terminology; but, without getting too technical or "in the weeds", it might also be worth considering that we could also contrast what the difference is between a "function" in a programming language versus a "function" in mathematics...  
    
</details><br><br>

***Don't forget to ask for summaries of your ChatBot session(s) and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatGPT)!***<br><br>\

(link in postlecture hw)

<details class="details-example"><summary style="color:blue"><u>Continue now...?</u></summary>

### Prelecture VS Postlecture HW
    
Feel free to work on the "Postlecture" HW below if you're making good progress and want to continue: in this case this is particularly reasonable as questions "6" and "7" below directly follow up and extend the "Prelecture" HW questions

*The benefits of continue would are that (a) it might be fun to try to tackle the challenge of working through some problems without additional preparation or guidance; and (b) this is a very valable skill to be comfortable with; and (c) it will let you build experience interacting with ChatBots (and beginning to understand their strengths and limitations in this regard)... it's good to have sense of when using a ChatBot is the best way to figure something out, or if another approach (such as course provided resources or a plain old websearch for the right resourse) would be more effective*
    
</details>    

## "Postlecture" HW [*submission along with "Prelecture" HW is due prior to next TUT*]

#### 6. The `df.describe()` method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> The answers here actually make it obvious why these can only be calculated for numeric variables in a dataset, which should help explain the answer to "4(a)" and "4(b)" above
>   
> Also notice that when `df.describe()` is used missing values are not explicitly removed, but `df.describe()`  provides answers anyway. Is it clear what `df.describe()` does with the data in each columns it analyzes if there is missing data in the column in question? 
>
> The next questions addresses removing rows or columns from a dataset in order to explicitly remove the presense of any missingness in the dataset (assuming we're not going to fill in any missing data values using any missing data imputation methods, which are beyond the scope of STA130); so, the behavior of `df.describe()` hints that explicitly removing missing may not always be necessary; but, the concern, though, is that not all methods may be able to handle missing data the way `df.describe()` does...
    
</details>


count: Number of non-null observations
mean: Average value
std: Standard deviation (spread of the data)
min: Minimum value
25%: 1st quartile (lower quartile, 25th percentile)
50%: 2nd quartile (median, 50th percentile)
75%:3rd quartile (upper quartile, 75th percentile)
max: Maximum value

NaN values are ignored when df.describe() is used. The data used for the above information are only NaN values. 

#### 7. Missing data can be considered "across rows" or "down columns".  Consider how `df.dropna()` or `del df['col']` should be applied to most efficiently use the available non-missing data in your dataset and  briefly answer the following questions in your own words.

1. Provide an example of a "use case" in which using `df.dropna()` might be peferred over using `del df['col']`<br><br>
    
2. Provide an example of "the opposite use case" in which using `del df['col']` might be preferred over using `df.dropna()` <br><br>
    
3. Discuss why applying `del df['col']` before `df.dropna()` when both are used together could be important<br><br>
    
4. Remove all missing data from one of the datasets you're considering using some combination of `del df['col']` and/or `df.dropna()` and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.<br><br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>


Notes:
df.dropna(): drops rows or columns with missing (NaN) values.
- df.dropna() -> drops rows with missing values
- df.dropna(subset=['col1']) -> drops rows with missing values from a specific column (ie. col1)
- df.dropna(axis=1) -> drops columns (variables) containing missing values.

def df['col']: deleting a column 

1. A use case where a df.dropna() is preferred over del df['col'] is when only a small amount of rows have missing values, such that when removed they won't significantly affect your dataset. When used on columns, it is preferred when the column contains a sparse amount of missing values.

2. A use case when del df['col'] is preferred over df.dropna() is when a column is irrelevant to your analysis, regardless of whether they have a lot of missing values or not. Removing the column will keep your other data (ie. other rows) intact.


3. This can be important because df.dropna() removes all rows with missing values. If you use del df['col'] on a column with a lot of missing values then apply df.dropna(), a lot more rows can be preserved since the column doesn't exist anymore. For example, there are a lot of missing values in the column 'type2' in my dataset; if i delete that column, a lot of my pokemons can stay in the dataframe.

4. [code below] 

In [8]:
# as shown above, all my missing data are in the column "Type 2". 
# If I use df.dropna() directly, all rows with an empty value in "Type 2" is removed.
# Results with df.dropna(): 414 rows x 12 columns — number of rows are nearly cut in half and no columns are removed.

# However, if i use del df['Type 2'] first then df.dropna(), as shown below,
# I can see that the column "Type 2" is removed and the dataframe has 800 rows x 11 columns. all of its rows are preserved. 

del df['Type 2']
df.dropna()
print(df)

       #                   Name   Type 1  HP  Attack  Defense  Sp. Atk  \
0      1              Bulbasaur    Grass  45      49       49       65   
1      2                Ivysaur    Grass  60      62       63       80   
2      3               Venusaur    Grass  80      82       83      100   
3      3  VenusaurMega Venusaur    Grass  80     100      123      122   
4      4             Charmander     Fire  39      52       43       60   
..   ...                    ...      ...  ..     ...      ...      ...   
795  719                Diancie     Rock  50     100      150      100   
796  719    DiancieMega Diancie     Rock  50     160      110      160   
797  720    HoopaHoopa Confined  Psychic  80     110       60      150   
798  720     HoopaHoopa Unbound  Psychic  80     160       60      170   
799  721              Volcanion     Fire  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
0         65     45           1      False  
1         80     60  


> Start a new ChatBot session **[but remember to first ask your ChatBot for summaries of your current session and perhaps coding results (so you can supply these in the homework as requested)]**, since your last ChatBot session has likely gotten quite long and has covered a lot of material at this point 
> - It can sometimes be helpful to reset ChatBot sessions to refocus them on the topics of inquiry without too much backlog history that might unintentionally bias things in certain directions and, of course, you can always re-introduce material from earlier conversations as it's relevant, such as for answering "D" based on reintroducing and updating code you made in a previous ChatBot session.  
> 

 link to first chatgpt conversation: https://chatgpt.com/share/4f0100d5-e698-41b4-babf-dd584d50a2a3\
 
 Here's a summary of our conversation:

DataFrame Overview:

You inquired about various ways to interact with a pandas DataFrame, including determining its dimensions using df.shape and the difference between attributes (e.g., df.shape) and methods (e.g., df.describe()).
Summarizing Data:

We discussed how to provide summaries for columns in a dataset, particularly using methods like df.describe() for numerical data and df['column'].value_counts() for categorical data.
Handling Missing Data:

We explored how to check for missing data in a DataFrame using methods like df.isnull().sum() and df.dropna().
We also discussed the behavior of df.describe() when dealing with missing data, particularly how it ignores NaN values in its calculations while showing the count of non-missing data.
Dropping Columns and Rows:

You learned how to use df.dropna() to remove rows or columns with missing values and how del df['col'] can be used to delete specific columns.
We emphasized the importance of using del df['col'] before df.dropna() when you want to remove columns with many missing values to avoid unnecessarily dropping rows with valid data in other columns.
Example and Error Handling:

When you attempted to delete a column using del df['Type 2'], an error occurred because the DataFrame (df) was not defined. We resolved this by first loading the dataset into df and then successfully deleting the column.
Overall, the discussion focused on best practices for cleaning and analyzing data using pandas, including handling missing data, summarizing data, and optimizing the use of both attributes and methods.

> #### ChatBot Scope Guidance
> 
> - This question is not interested in the general benefits of imputing missing data, or the general benefits of using `df.dropna()` and/or `del df['col']` to remove missing data, just how to most efficiently remove missing data if a user chooses to do so
> 
> - More sophisticated analyses for "filling in" rather than removing missing data (as considered here) are possible (based on making assumptions about missing data and using specific imputation methods or models) but these are "beyond the scope" of this homework assignment so this topics can be safely ignored for now
> 
> #### ChatBot Code Troubleshooting
> 
> A key issue to be aware of when asking ChatBots for help with something is that they are not running and checking code for correctess, and they often intertwine written instructions with code instructions; so, BEFORE YOU RUN ANY CODE provided by a ChatBot, you should check the following
> 
> 1. If this code changes an object or data, are you sure you want to run this code?
> 2. Can you easily "undo" the results of running code (e.g., from a copy `df_saved=df.copy()` or reloading the data) if running the code doesn't do what you want?
> 3. Is the state of the data what is expected by the code? Or have the objects been updated and changed so they're no longer what the code expects them to be? 
> 
> #### If you get any `Python` errors, copy and paste them into the ChatBot prompt and see if it can help you resove them; but, keep in mind the final point above becasue the ChatBot might not be aware of the state of your objects relative to the code it's producing...

</details><br>

#### 8. Give brief explanations in your own words for any requested answers to the questions below

> This problem will guide you through exploring how to use a ChatBot to troubleshoot code using the "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" data set 
> 
> To initialially constrain the scope of the reponses from your ChatBot, start a new ChatBot session with the following slight variation on the initial prompting approach from "2" above
> - "I am going to do some initial simple summary analyses on the titanic data set I've downloaded (https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv) which has some missing values, and I'd like to get your help understanding the code I'm using and the analysis it's performing"
        
1. Use your ChatBot session to understand what `df.groupby("col1")["col2"].describe()` does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you
    
> If needed, you can help guide the ChatBot by showing it the code you've used to download the data **AND provide it with the names of the columns** using either a summary of the data with `df.describe()` or just `df.columns` as demonstrated [here](../CHATLOG/COP/00017_copilot_groupby.md)
    


In [9]:
#set up data
url_titanic = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df_titanic = pd.read_csv(url_titanic)
df_titanic

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

df.groupby("col1")["col2"].describe() groups the observations in column 1, then analyzes the data from column 2 based on the "groups" from column 1. This includes the count (number of repeated observations in column 1), and the mean, std, etc from their corresponding values in column 2. 

for example:

In [10]:
df_titanic.groupby("sex")["age"].describe()
#here, i am able to find out statistical information about the titanic victims sorted by gender.

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

2. Assuming you've not yet removed missing values in the manner of question "7" above, `df.describe()` would have different values in the `count` value for different data columns depending on the missingness present in the original data.  Why do these capture something fundamentally different from the values in the `count` that result from doing something like `df.groupby("col1")["col2"].describe()`?

> Questions "4" and "6" above address how missing values are handled by `df.describe()` (which is reflected in the `count` output of this method); but, `count` in conjunction with `group_by` has another primary function that's more important than addressing missing values (although missing data could still play a role here).


df.describe() and df.groupby("col1")["col2"].describe() provides different values because they handle missing values differently. df.describe() gathers information about all rows and columns regardless of whether they have missing values or not. The "count" in this case includes those with missing values.

meanwhile, df.groupby("col1")["col2"].describe() first groups repeating values in col1, then gathers information about the values in col2 within each group. If there are missing values in col1, rows with missing values are excluded from grouping, which is why the count may be lower. The "count" value does not include missing values in col2 as well. 

3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT<br><br>
    
    1. Forget to include `import pandas as pd` in your code 
       <br> 
       Use Kernel->Restart from the notebook menu to restart the jupyter notebook session unload imported libraries and start over so you can create this error
       <br><br>
       When python has an error, it sometimes provides a lot of "stack trace" output, but that's not usually very important for troubleshooting. For this problem for example, all you need to share with ChatGPT or search on google is `"NameError: name 'pd' is not defined"`<br><br>


Chatgpt is better than google because it is more case-specific and identifies the problem immediately when i copy and pasted the error message into the chatbox. It also offered a direct solution to the issue. Compared to google, i had to browse a bit to figure out the solution. 

    B. Mistype "titanic.csv" as "titanics.csv"
       <br> 
       If ChatBot troubleshooting is based on downloading the file, just replace the whole url with "titanics.csv" and try to troubleshoot the subsequent `FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'` (assuming the file is indeed not present)
       <br><br>
       Explore introducing typos into a couple other parts of the url and note the slightly different errors this produces<br><br>

Chatgpt is better. The error message was really long and complicated. When i copy and pasted it into chatgpt, it explained that it was because the URL doesn't exist, and was able to correct me by finding where i added the extra 's'. compared to using google, it wouldn't have been able to help me identify where exactly the typo occured, and I would've had to spend some time reading over every character of the URL to see where the typo is. 

    C. Try to use a dataframe before it's been assigned into the variable
       <br> 
       You can simulate this by just misnaming the variable. For example, if you should write `df.groupby("col1")["col2"].describe()` based on how you loaded the data, then instead write `DF.groupby("col1")["col2"].describe()`
       <br><br>
       Make sure you've fixed your file name so that's not the error any more<br><br>

Chatbots are better. once again, i copy and pasted the error message into the chatbox, and it identified the problem immediately. while a simple google search also answered my question, chatgpt provided an alternative solution (the same as my original code before removing the variable) for me to copy and paste.

  
    D. Forget one of the parentheses somewhere the code
       <br>
       For example, if the code should be `pd.read_csv(url)` the change it to `pd.read_csv(url`<br><br>

both are fine, since it is quite easy to fix. the error message identified the missing parenthesis and provided the location so it was unnecessary to use either chatgpt or google. When pasted into chatgpt, it was also able to help fix the problem.

     
    E. Mistype one of the names of the chained functions with the code 
       <br>
       For example, try something like `df.group_by("col1")["col2"].describe()` and `df.groupby("col1")["col2"].describle()`<br><br>
        

there is an attribute error message. 

chatbots are better. when pasted into chatgpt, it identified the issue and fixed it for me. googling it gave me the correct function spelling as well, but there are no explanation and it was not specific like chatbots are.

   F. Use a column name that's not in your data for the `groupby` and column selection 
       <br>
       For example, try capitalizing the columns for example replacing "sex" with "Sex" in `titanic_df.groupby("sex")["age"].describe()`, and then instead introducing the same error of "age"<br><br>
        

chatgpt is better. when i pasted the error message into the chatbox, it identified the problem and gave me the right names for the col1 and col2 values in titanic_df.groupby("col1")["col2"].describe(). google is not able to do the same thing and it can't identify the correct name i should replace the error with. 

G. Forget to put the column name as a string in quotes for the groupby and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question


For example, something like `titanic_df.groupby(sex)["age"].describe()`, and then `titanic_df.groupby("sex")[age].describe()`


chatbots are better. on jupyter, the error told me that "NameError: name 'sex' is not defined". pasting the entire error message into chat was able to identify what exactly i should do to fix the error. it would be hard to google about what issue if i didn't know it was because i didn't add it as a string.

#### 9. Have you reviewed the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?<br>
    
<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Just answering "Yes" or "No" or "Somewhat" or "Mostly" or whatever here is fine as this question isn't a part of the rubric; but, the midterm and final exams may ask questions that are based on the tutorial and lecture materials; and, your own skills will be limited by your familiarity with these materials (which will determine your ability to actually do actual things effectively with these skills... like the course project...)
    
</details>

Yes


    
***Don't forget to ask for summaries of your ChatBot session(s) and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatGPT)!***



link to second chatgpt conversation: https://chatgpt.com/share/2c4e6665-be1a-418b-9cd0-4c524aa79fbe

Here’s a summary of our conversation:

Loading the Titanic Dataset: You initially encountered issues loading the Titanic dataset and had a typo in the URL, which caused an HTTP 404 error. We corrected it to the proper URL: "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv".

Pandas Library Import: You also ran into a NameError because you hadn't imported the pandas library (import pandas as pd). Once this was added, your code executed correctly.

Using df.groupby(): You wanted to group data and calculate descriptive statistics using df.groupby("col1")["col2"].describe(). At first, you mistakenly used group_by() (with an underscore), but the correct method is groupby() (without the underscore).

Dealing with Missing Values: We discussed how missing values affect the count when using df.describe() versus df.groupby("col1")["col2"].describe(). In groupby(), missing values in either column can reduce the count, but df.describe() only counts missing values in individual columns.

KeyError with Case-Sensitive Columns: You encountered a KeyError when trying to group by "Sex", which was due to a case-sensitive column name. The correct column name in the dataset is "sex" (lowercase), and once this was fixed, the code worked.

Let me know if you need further assistance!

# Recommended Additional Useful Activities [Optional]

The "Ethical Profesionalism Considerations" and "Current Course Project Capability Level" sections below **are not a part of the required homework assignment**; rather, they are regular weekly guides covering (a) relevant considerations regarding professional and ethical conduct, and (b) the analysis steps for the STA130 course project that are feasible at the current stage of the course<br><br>

<details class="details-example"><summary style="color:blue"><u>Ethical Professionalism Considerations</u></summary>

## Ethical Professionalism Considerations

> If the observed data is "no events occured" does this mean the data is "missing" and [should be ignored](https://priceonomics.com/the-space-shuttle-challenger-explosion-and-the-o)?
> 
> - NASA: \<determines temperature doesn't affects "o-ring" by subseting data to just "o-ring" incidents\>
> - Also NASA: \<launches the shuttle on a cold day\>

|No apparent "o-ring" failure and temperature relationship|Apparent between "o-ring" failure and temperature relationship|
|:-|:-|
if you just look at "o-ring" failure event data|if you instead look at ALL the data as you should|
|![](https://etzq49yfnmd.exactdn.com/wp-content/uploads/2022/03/image06-14.png)|![](https://etzq49yfnmd.exactdn.com/wp-content/uploads/2022/03/image02-33.png)|
|![](https://upload.wikimedia.org/wikipedia/commons/8/8b/Shuttle_Challenger_explosion.gif?20190203170223)|![](https://i.makeagif.com/media/10-04-2014/nT57xW.gif)|

<br>
    
</details>    

<details class="details-example"><summary style="color:blue"><u>Current Course Project Capability Level</u></summary>

## Current Course Project Capability Level

> The data we'll use for the STA130 course project is based on the [Canadian Social Connection Survey](https://casch.org/cscs). Please see the [data use agreement](https://static1.squarespace.com/static/60283c2e174c122f8ebe0f39/t/6239c284d610f76fed5a2e69/1647952517436/Data+Use+Agreement+for+the+Canadian+Social+Connection+Survey.pdf) regarding the appropriate and ethical professional use of this data (available at the bottom of the [CSCS](https://casch.org/cscs) webpage).
> 
> 1. Have a very quick look at the list of available variables available using the [link](https://drive.google.com/file/d/1ISVymGn-WR1lcRs4psIym2N3or5onNBi/view) (again at the bottom of the [CSCS](https://casch.org/cscs) webpage); then, 
> 2. examine the code in the first thirteen code cells of [STA130F24_CourseProject.ipynb](https://github.com/pointOfive/stat130chat130/blob/main/CP/STA130F24_CourseProject.ipynb) to get an initital understanding of how we might subset to different studies included in the [data](https://drive.google.com/file/d/1mbUQlMTrNYA7Ly5eImVRBn16Ehy9Lggo/view) (again accessible at the bottom of the [CSCS](https://casch.org/cscs) webpage); then,     
> 3. review the fourteenth and fifteenth cells (with the comments "Here's a high level summary of the data" and "And here are some explanations about the columns in the data") a little more closely to get a better sense of which columns seem to be the most interesting and whether or not they seem to have a lot of missing data
    
</details>        

### Afterward

Here are few ideas of some other kinds of interactions you might consider exploring with a ChatBot...

> While these are likely to be extremely practically valuable, they are not a part of the homework assignment, so do not include anything related to these in your homework submission

- With respect to improving ones ability in statistics, coding, communication, and other key data science skills
    - what is the ChatBots perception its own capabilities and uses as an AI-driven assistance tool 
    - and does ChatBots assessment of itself influence or agree with your own evalution of the ChatBot? 

- ChatBots can introduce and explain the "World War 2 planes" problem and the "Monte Hall" problem... 
    - how well does do they seem to do and introducing and explaining other "unintuitive surprising statistics paradoxes"?

- If you consider the process of writing about why you chose to take this course, and the skills you were hoping to build through this course with respect to your current ideas about what possible careers 
    - and how do you think the exercise would be different if you framed it as a dialogue with a ChatBot
    - and do you think the difference could be positive and productive, or potentially biasing and distracting?
    
- ChatBots sometimes immediately responds in simple helpful ways, but other times it gives a lot of extraneous information that can be overwheling... are you able to prompt and interact with ChatBots in manner that keeps its reponses helpful and focused on what you're interested in? 

- ChatBots tends to respond in a fairly empathetic and supportive tone...
    - do you find it helpful to discuss concerns you might have about succeeding in the course (or entering university more generally) with a ChatBot?
    
- For what purposes and in what contexts do you think a ChatBot could provide suggestions or feedback about your experiences that might be useful? 
